include/crypto/ml_kem_types.h

```diff
@@ -0,0 +1,41 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#ifndef ML_KEM_TYPES_H
+# define ML_KEM_TYPES_H
+# pragma once
+
+# ifndef OPENSSL_NO_ML_KEM
+
+typedef struct ossl_ml_kem_scalar_st ossl_ml_kem_scalar;
```

Comment:
We are using this kind of name in most other places..
ML_KEM_SCALAR

Reply:
Note I used POLY as the name, (yes its a scalar, but polynomial is what these scalars actually are).

Reply:
FWIW, the type is internal, its name is not visible in public header files, but sure, we should try to be reasonably consistent.

From a mathematical perspective, the fact that we're notionally in a polynomial ring is only one "up-to-isorphism" perspective.
- We start with a prime field: $`F_q = \mathbb{Z}_q`$
    * where $q = 256 \cdot 13 + 1$
    * As with every prime field, its multiplicative group is cyclic of order $`q - 1 = 256 \cdot 13`$
    * So there must be an element whose multiplicative order is $256$ exactly, i.e. a primitive $256^{\mathtt{th}}$ root of unity
    * There are no elements $r$, with $r^{256} + 1 = 0$, since then $r^{q-1} = -1^{13} = -1$, instead of $1$.
    * One of these happens to be $\zeta = 17$.
    * None of the odd powers of $\zeta$ have square roots in $F_q$
        - They are all also primitive $`256^{\mathtt{th}}`$ roots, but $17$ is chosen as the smallest representative.
- Now, we consider the polynomial ring: $`R_q = \mathbb{Z}_q[x] / (x^{256} + 1)`$
- Being a ring (almost like a field, just no division), we can build *modules* over $R_q$.
    * Modules are  essentially analogues of vector spaces from ~first year uni math class.
    * Especially, when their elements are just column "vectors" of ring elements, added and scaled coördinate-wise.
- However, our ring turns out to be isomorphic to a direct sum of much simpler rings:
    * $`R_q \cong \bigoplus_{i=0}^{127} \left(\mathbb{Z}_q[x] / (x^2 + \zeta^{2i+1})\right)`$
    * The isomorphism is given by an analogue of the iFFT (integer Fast Fourier Transform)
    * This turns polynomial multiplication (convolutions over the lists of coëfficients into pair-wise multiplications, making computations in the ring dramatically more efficient.
    * The parts of the direct sum look a lot like complex numbers with real and imaginary parts in $F_q$
        - When an element $r$ has no square roots in a field $F$,
        - For any non-square $r$, the ring $F[x]/(x^2 + r)$ consists of just linear terms of the form $a + b\cdot x$, with $a, b \in F_q$.
        - Their sums are trivially coëfficient-wise
        - But their product $(a + bx)(c + dx) \cong ac + bd\cdot x^2 + (ac + bd) x$:
            - simplifies to $(ac - r\cdot bd) + (ac + bd)x$ in close analogy to complex numbers (where $r = 1$)
            - because (x^2 + r) is the zero of the quotient ring, so $`x^2 \cong -r`$
- So, up to nomenclature (isomorphism), our actual ring (a "scalar") is just an array of 128 funny complex numbers added and multiplied coördinate-wise, where the square of the imaginary part is some odd power of $\zeta$, rather than $-1$.
    * And our *module* (vectors over the ring) is then just arrays of these arrays.
    * And our matrices are arrays of vectors (one per row) so are really three dimensional over these "complex" numbers.

So when all is said done, we can don't think of think of these as polynomials, they aren't used directly as such. That largely just motivates the construction. They're never used as functions of their variable, for example.

However, while most of the computations happen on the "Fourier" (list of "complex numbers") side of the isomorphism, the "errors" in LWE (learning with errors) are added on the "polynomial" (inverse NTT) side. This likely helps to diffuse the noise throughout the values computed on the Forier (NTT) side (I'll leave the details of why to the experts).

Bottom line, ones does not have to think of these as "polynomials", unless that makes it much easier to compare the code with the algorithms in FIPS 203/204. Just using the term "scalar" in (element of $R_q$ with 256 coëfficients from $F_q$, or its Fourier transform with 128 pairs of coëfficients from $F_q$) avoids the topic of what these arrays of numbers mean, or which side of the NTT isomorphism is the one true representation.

Reply:
regardless you still access them at some point as a ->c... (which is a coefficient).

Reply:
Yes, but that access is representation agnostic, we call them scalars on both the "Polynomial" and the "NTT" side, and use the same data type for both. One could perhaps argue that these could be stored identically but given distinct types to avoid mixing oil and water. But that was not done in the BoringSSL code, and I don't think it is compelling, it would clutter the code.